server/src/main/java/org/elasticsearch/index/shard/IndexWriteLoad.java

```diff
@@ -0,0 +1,217 @@
+        for (IndexShardStats indexShardsStats : indexShards.values()) {
+            final var shardStats = Arrays.stream(indexShardsStats.getShards())
+                .filter(stats -> stats.getShardRouting().primary())
+                .findFirst()
+                // Fallback to a replica if for some reason we couldn't find the primary stats
+                .orElse(indexShardsStats.getAt(0));
+            final var indexingShardStats = shardStats.getStats().getIndexing().getTotal();
+            indexWriteLoadBuilder.withShardWriteLoad(
+                shardStats.getShardRouting().id(),
+                indexingShardStats.getWriteLoad(),
+                indexingShardStats.getTotalActiveTimeInMillis()
+            );
+        }
+        return indexWriteLoadBuilder.build();
+    }
+
+    private final double[] shardWriteLoadLoad;
+    private final long[] shardUptimeInMillis;
+
+    private IndexWriteLoad(double[] shardWriteLoadLoad, long[] shardUptimeInMillis) {
+        assert shardWriteLoadLoad.length == shardUptimeInMillis.length;
+        this.shardWriteLoadLoad = shardWriteLoadLoad;
+        this.shardUptimeInMillis = shardUptimeInMillis;
+    }
+
+    public IndexWriteLoad(StreamInput in) throws IOException {
+        this(in.readDoubleArray(), in.readLongArray());
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeDoubleArray(shardWriteLoadLoad);
+        out.writeLongArray(shardUptimeInMillis);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field(SHARDS_WRITE_LOAD_FIELD.getPreferredName(), shardWriteLoadLoad);
+        builder.field(SHARDS_UPTIME_IN_MILLIS.getPreferredName(), shardUptimeInMillis);
+        return builder;
+    }
+
+    public static IndexWriteLoad fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    @Nullable
+    public Double getWriteLoadForShard(int shardId) {
+        return getWriteLoadForShard(shardId, null);
+    }
+
+    public Double getWriteLoadForShard(int shardId, Double defaultValue) {
+        checkBounds(shardId);
+
+        double load = shardWriteLoadLoad[shardId];
+        if (load == UNKNOWN_LOAD) {
+            return defaultValue;
+        } else {
+            return load;
+        }
+    }
+
+    @Nullable
+    public Long getUptimeInMillisForShard(int shardId) {
+        return getUptimeInMillisForShard(shardId, null);
+    }
+
+    public Long getUptimeInMillisForShard(int shardId, Long defaultValue) {
+        checkBounds(shardId);
+
+        long uptime = shardUptimeInMillis[shardId];
+        if (uptime == UNKNOWN_UPTIME) {
+            return defaultValue;
+        } else {
+            return uptime;
+        }
+    }
+
+    // Visible for testing
+    public int numberOfShards() {
+        return shardWriteLoadLoad.length;
+    }
+
+    private void checkBounds(int shardId) {
```

Comment:
I wonder if we need this? We will get it from the array anyway - and it is also sort of assertion territory, i.e., at bug if we hit this?